pkg/backends/influxdb/flush.go

```diff
@@ -0,0 +1,256 @@
+	writeName(f.writer, name, tags)
+	_, _ = f.writer.Write([]byte(fmt.Sprintf("value=%g %d\n", value, f.timestampSeconds)))
+	f.metricCount++
+	f.maybeFlush()
+}
+
+func (f *flush) addSet(name string, tags gostatsd.Tags, value uint64) {
+	writeName(f.writer, name, tags)
+	_, _ = f.writer.Write([]byte(fmt.Sprintf("count=%d %d\n", value, f.timestampSeconds)))
+	f.metricCount++
+	f.maybeFlush()
+}
+
+func (f *flush) addBaseTimer(name string, timer gostatsd.Timer) {
+	var sb strings.Builder
+	if !f.disabledSubtypes.Lower {
+		sb.WriteString(fmt.Sprintf("lower=%g,", timer.Min))
+	}
+	if !f.disabledSubtypes.Upper {
+		sb.WriteString(fmt.Sprintf("upper=%g,", timer.Max))
+	}
+	if !f.disabledSubtypes.Count {
+		sb.WriteString(fmt.Sprintf("count=%d,", timer.Count))
+	}
+	if !f.disabledSubtypes.CountPerSecond {
+		sb.WriteString(fmt.Sprintf("rate=%g,", timer.PerSecond))
+	}
+	if !f.disabledSubtypes.Mean {
+		sb.WriteString(fmt.Sprintf("mean=%g,", timer.Mean))
+	}
+	if !f.disabledSubtypes.Median {
+		sb.WriteString(fmt.Sprintf("median=%g,", timer.Median))
+	}
+	if !f.disabledSubtypes.StdDev {
+		sb.WriteString(fmt.Sprintf("stddev=%g,", timer.StdDev))
+	}
+	if !f.disabledSubtypes.Sum {
+		sb.WriteString(fmt.Sprintf("sum=%g,", timer.Sum))
+	}
+	if !f.disabledSubtypes.SumSquares {
+		sb.WriteString(fmt.Sprintf("sum_squares=%g,", timer.SumSquares))
+	}
+	for _, pct := range timer.Percentiles {
+		sb.WriteString(fmt.Sprintf("%s=%g,", pct.Str, pct.Float))
+	}
+	if sb.Len() == 0 { // wat
+		return
+	}
+	writeName(f.writer, name, timer.Tags)
+	buf := sb.String()
+	_, _ = f.writer.Write([]byte(fmt.Sprintf("%s %d\n", buf[:len(buf)-1], f.timestampSeconds)))
+	f.metricCount++
+	f.maybeFlush()
+}
+
+func (f *flush) addHistogramTimer(name string, timer gostatsd.Timer) {
+	writeName(f.writer, name, timer.Tags)
+
+	var sb strings.Builder
+	for histogramThreshold, count := range timer.Histogram {
+		bucket := "le.+Inf"
+		if !math.IsInf(float64(histogramThreshold), 1) {
+			bucket = "le." + strconv.FormatFloat(float64(histogramThreshold), 'f', -1, 64)
+		}
+		sb.WriteString(bucket)
+		sb.WriteByte('=')
+		sb.WriteString(strconv.Itoa(count))
+		sb.WriteByte(',')
+	}
+	buf := sb.String()
+	_, _ = f.writer.Write([]byte(fmt.Sprintf("%s %d\n", buf[:len(buf)-1], f.timestampSeconds)))
+	f.metricCount++
+	f.maybeFlush()
+}
+
+func (f *flush) maybeFlush() {
+	if f.metricCount >= f.metricsPerBatch {
+		f.flush()
+	}
+}
+
+func (f *flush) finish() {
+	if f.metricCount > 0 {
+		f.flush()
+	}
+	f.releaseBuffer(f.buffer)
+}
+
+func (f *flush) flush() {
+	// I don't believe that a gzip.Writer sending to a bytes.Buffer can actually fail, but the
+	// flate.huffmanBitWriter at the core of the object graph *does* have an internal error it
+	// can theoretically produce, so we won't ignore the error.
+	err := f.writer.Close()
+	if err == nil {
+		f.cb(f.buffer, f.metricCount)
+	} else {
+		atomic.AddUint64(f.errorCounter, 1)
+		f.releaseBuffer(f.buffer)
```

Comment:
Is the release buffer a paradigm we have used throughout the project?

Reply:
Eh... not as explicit as this.  We do have the "a semaphore is a channel of buffers" though.

I'm not a huge fan of this honestly, but it was the cleanest way I found to keep `processMetrics` small-ish.